### ISSUE TITLE
[Feat] Enable LoRA adaptation on Qwen3-0.6B

### ISSUE THREAD

**Issue by EunjuYang** [nnstreamer/nntrainer#3831]
We need to enable LoRA adaptation on Qwen3-0.6B.
To this end, we set two sub-milestones:
   1) Train Qwen3-0.6B with FP32
   2) Train Qwen3-0.6B with 4-bit weight

# Train Qwen3-0.6B with FP32
**GOAL**: we need to finetune LoRA (only) with FP32

## What we've done
- [x] LoRA implementation (`forwarding`, `calcDerivative`, `calcGradient`)
- [x] Qwen3-0.6B FP32 Inference (Applications/CausalLM)

## Works to do
- Implement `calcDerivative` of the following layers: (Please implement `calcDerivative` and **unittest** for the layer)
    - [ ] CausalLM/layers/lm_head
    - [ ] CausalLM/mha_core.cpp
    - [ ] #3836
    - [ ] CausalLM/rms_norm
    - [ ] CausalLM/tie_word_embedding

- Test `LoRA` adaptation with incremental_fowarding
    - [ ] stacking FC layers and implement training application for LoRA adaptation (verify whether it works)

**Comment by baek2sm** [comment]
reference code (#3832) for this, but thorough verification is required.

**Comment by EunjuYang** [comment]
# (Phase 1) Initiate training CausalLM with batch 1 training

Currently, CausalLM has been implemented with functionality to enable inference through incremental_forwarding. However, in the case of the current train() function, since it follows a structure that calls forwarding()->backwarding(), we need to implement the forwarding() accordingly.
But forwarding() has a limitation in that it cannot handle variable-length inputs. Therefore, ultimately, we need to enable training by providing masks as inputs when multiple variable-length inputs are given within a single batch. Prior to achieving this ultimate goal, we first need to implement training functionality for batch size = 1. Thus, our immediate objective is to complete the implementation of a training method where, for batch size = 1, the input dimensions are adjusted individually for each input as follows:


(1) Batch Size = 1

(2) Implement Data Loader for varient length inputs **(updated)**
=> zero padding (masking)

(3) Within each layer, assume that the input passed to forwarding() always matches the actual input length; call incremental_forwarding(from=0, to=input height length)
```
forwarding(nntrainer::RunLayerContext &context, bool training){

   auto input_dim = context.getInputDimensions()[0];
   auto to = input_dim.height();
   incremental_forwarding(context, 0, to);
}
```

By doing so, we expect that the existing train() structure with forwarding()->backwarding() can be used as-is.


------------


If we follow the plan above, what we have to do is
- Update layers in Applications/CausalLM (define forwarding like (3))
- Implement calcDerivative of layers in Applications/CausalLM 
- Implement DataLoader to support the case which lenghts are varying.
